src/lib/storage/dictionary_segment.hpp

```diff
@@ -25,55 +30,128 @@ class DictionarySegment : public BaseSegment {
 
   // returns an underlying dictionary
-  std::shared_ptr<const std::vector<T>> dictionary() const;
+  std::shared_ptr<const std::vector<T>> dictionary() const { return _dictionary; }
 
   // returns an underlying data structure
-  std::shared_ptr<const BaseAttributeVector> attribute_vector() const;
+  std::shared_ptr<const BaseAttributeVector> attribute_vector() const { return _attribute_vector; }
 
   // return the value represented by a given ValueID
-  const T& value_by_value_id(ValueID value_id) const;
+  const T& value_by_value_id(ValueID value_id) const { return _dictionary.at(value_id); }
 
   // returns the first value ID that refers to a value >= the search value
   // returns INVALID_VALUE_ID if all values are smaller than the search value
-  ValueID lower_bound(T value) const;
+  ValueID lower_bound(T value) const {
+    const auto iterator = std::lower_bound(_dictionary->begin(), _dictionary->end(), value);
+    if (iterator == _dictionary->end()) {
+      return INVALID_VALUE_ID;
+    } else {
+      return ValueID(std::distance(_dictionary->begin(), iterator));
+    }
+  }
 
   // same as lower_bound(T), but accepts an AllTypeVariant
-  ValueID lower_bound(const AllTypeVariant& value) const;
+  ValueID lower_bound(const AllTypeVariant& value) const { return lower_bound(type_cast<T>(value)); }
 
   // returns the first value ID that refers to a value > the search value
   // returns INVALID_VALUE_ID if all values are smaller than or equal to the search value
-  ValueID upper_bound(T value) const;
+  ValueID upper_bound(T value) const {
+    const auto iterator = std::upper_bound(_dictionary->begin(), _dictionary->end(), value);
+    if (iterator == _dictionary->end()) {
+      return INVALID_VALUE_ID;
+    } else {
+      return ValueID(std::distance(_dictionary->begin(), iterator));
+    }
+  }
 
   // same as upper_bound(T), but accepts an AllTypeVariant
-  ValueID upper_bound(const AllTypeVariant& value) const;
+  ValueID upper_bound(const AllTypeVariant& value) const { return upper_bound(type_cast<T>(value)); }
 
   // return the number of unique_values (dictionary entries)
-  size_t unique_values_count() const;
+  size_t unique_values_count() const { return _dictionary->size(); }
 
   // return the number of entries
-  size_t size() const override;
+  size_t size() const override { return _attribute_vector->size(); }
 
   // returns the calculated memory usage
-  size_t estimate_memory_usage() const final;
+  size_t estimate_memory_usage() const final {
+    const size_t dictionary_size = _dictionary->capacity() * sizeof(T);
+    return dictionary_size + _attribute_vector->estimate_memory_usage();
+  }
 
  protected:
   std::shared_ptr<std::vector<T>> _dictionary;
   std::shared_ptr<BaseAttributeVector> _attribute_vector;
+
+  void _compress_values(const std::vector<T>& column_values) {
+    std::vector<uint32_t> lookup_indices(column_values.size());
```

Comment:
Your code here has good performance, but understanding it is a bit hard. I think adding a comment that explains the basic idea what data structure is used to store what information at the beginning would be a good idea.